tcmur_cmd_handler.c

```diff
@@ -294,6 +298,657 @@ static int handle_write_verify(struct tcmu_device *dev, struct tcmulib_cmd *cmd)
+	 * From spc4r31, section 7.8.6.1 Device Identification VPD page
+	 * overview
+	 */
+	if ((tgt_desc[5] & 0x0f) != 0x3) {
+		tcmu_dev_err(udev, "Id target DESIGNATOR TYPE other than NAA not supported!\n");
+		return tcmu_set_sense_data(sense, ILLEGAL_REQUEST,
+					   ASC_INVALID_FIELD_IN_PARAMETER_LIST,
+					   NULL);
+	}
+	/*
+	 * Check for matching 16 byte length for NAA IEEE Registered Extended
+	 * Assigned designator
+	 */
+	if (tgt_desc[7] != 16) {
+		tcmu_dev_err(udev, "Id target DESIGNATOR LENGTH should be 16, but it's: %d\n",
+			     tgt_desc[7]);
+		return tcmu_set_sense_data(sense, ILLEGAL_REQUEST,
+					   ASC_INVALID_FIELD_IN_PARAMETER_LIST,
+					   NULL);
+	}
+
+	/*
+	 * Check for NAA IEEE Registered Extended Assigned header.
+	 */
+	if ((tgt_desc[8] >> 4) != 0x06) {
+		tcmu_dev_err(udev, "Id target NAA designator type: 0x%x\n",
+			     tgt_desc[8] >> 4);
+		return tcmu_set_sense_data(sense, ILLEGAL_REQUEST,
+					   ASC_INVALID_FIELD_IN_PARAMETER_LIST,
+					   NULL);
+	}
+
+	/*
+	 * Source designator matches the local device
+	 */
+	if (index == xcopy->stdi) {
+		memcpy(&xcopy->src_tid_wwn[0], &tgt_desc[8],
+		       XCOPY_NAA_IEEE_REGEX_LEN);
+
+		if (!memcmp(wwn, xcopy->src_tid_wwn, XCOPY_NAA_IEEE_REGEX_LEN))
+			xcopy->src_dev = udev;
+	}
+
+	/*
+	 * Destination designator matches the local device.
+	 */
+	if (index == xcopy->dtdi) {
+		memcpy(xcopy->dst_tid_wwn, &tgt_desc[8],
+		       XCOPY_NAA_IEEE_REGEX_LEN);
+
+		if (!memcmp(wwn, xcopy->dst_tid_wwn, XCOPY_NAA_IEEE_REGEX_LEN))
+			xcopy->dst_dev = udev;
+	}
+
+	return SAM_STAT_GOOD;
+}
+
+static int xcopy_parse_target_descs(struct tcmu_device *udev,
+				    struct xcopy *xcopy,
+				    uint8_t *tgt_desc,
+				    uint16_t tdll,
+				    uint8_t *sense)
+{
+	int i, ret;
+
+	if (tdll > RCR_OP_MAX_TARGET_DESC_COUNT * XCOPY_TARGET_DESC_LEN) {
+		tcmu_dev_err(udev, "Only %u target descriptor(s) supported, but there are %u\n",
+			     RCR_OP_MAX_TARGET_DESC_COUNT, tdll / XCOPY_TARGET_DESC_LEN);
+		return tcmu_set_sense_data(sense, ILLEGAL_REQUEST,
+					   ASC_INVALID_FIELD_IN_PARAMETER_LIST,
+					   NULL);
+	}
+
+	for (i = 0; i < RCR_OP_MAX_TARGET_DESC_COUNT; i++) {
+		/*
+		 * Only Identification Descriptor Target Descriptor support
+		 * for now.
+		 */
+		if (tgt_desc[0] == XCOPY_TARGET_DESC_TYPE_CODE_ID) {
+			ret = xcopy_parse_target_id(udev, xcopy, tgt_desc, i, sense);
+			if (ret != SAM_STAT_GOOD)
+				return ret;
+
+			tgt_desc += XCOPY_TARGET_DESC_LEN;
+		} else {
+			tcmu_dev_err(udev, "Unsupport target descriptor type code 0x%x\n",
+				     tgt_desc[0]);
+			return tcmu_set_sense_data(sense, ILLEGAL_REQUEST,
+						   ASC_UNSUPPORTED_TARGET_DESC_TYPE_CODE,
+						   NULL);
+		}
+	}
+
+	if (xcopy->src_dev)
+		ret = xcopy_locate_udev(udev->ctx, xcopy->dst_tid_wwn,
+					&xcopy->dst_dev);
+	else if (xcopy->dst_dev)
+		ret = xcopy_locate_udev(udev->ctx, xcopy->src_tid_wwn,
+					&xcopy->src_dev);
+	else
+		ret = SAM_STAT_CHECK_CONDITION;
```

Comment:
@mikechristie I think we can refer to spec4r31, section 6.3.7.1 Segment descriptors introduction:

"If the copy target device identified by a SOURCE TARGET DESCRIPTOR INDEX field or a DESTINATION TARGET DESCRIPTOR INDEX field is not accessible to the copy manager, then the command shall be terminated with CHECK CONDITION status, with the sense key set to COPY ABORTED, and the additional sense code set to UNREACHABLE COPY TARGET."